designs/default-zero-value.md

```diff
@@ -0,0 +1,888 @@
+(As of 2021-03-18, There are 4,805 structure members used in output marked as
+required, of which 458 target structures; there are 3,382 members used in input
+marked as required, of which 367 target structures).
+
+Assume that members that target structures followed the same nullability rules
+as other members, and consider the following model on _day 1_:
+
+```
+structure Foo {
+    // Nullable because Baz has no zero value due to a required member.
+    // Only members with a zero value can be non-null because we must
+    // support the ability to remove the required trait by replacing
+    // it with the @default trait.
+    @required
+    baz: Baz
+}
+
+structure Baz {
+    @required
+    bar: String
+}
+```
+
+Consider if the model was changed to this on _day 2_:
+
+```
+structure Foo {
+    // Non-nullable because Baz now has a zero value. A breaking change!
+    @required
+    baz: Baz
+}
+
+structure Baz {
+    @default
+    bar: String
+}
+```
+
+In this example, `Baz` evolved backward compatibly, but Foo did not because
+`Foo$baz` is now non-null.
+
+There are a few ways to address this:
+
+1. We don't ever treat required members that reference structures as non-null.
+   This is the option chosen for this proposal.
+
+2. We update `Foo$baz` to have a required trait like `@required(nullable: true)`
+   to force nullability. However, this won't work because the team that owns
+   `Baz` would need to coordinate their change with every team that references
+   their structure to let them know they need to add the `nullable` attribute to
+   required traits.
+
+```
+structure Foo {
+    @required(nullable: true)
+    baz: Baz
+}
+```
+
+3. When `Baz` is updated on _day 2_, the `@box` trait is added to `Baz` to
+   prevent any structure member from ever using a non-null accessor with it.
+   This would almost certainly be forgotten by modelers, and would all but
+   require smithy-diff to point out the compatibility issue.
+
+```
+@box
+structure Baz {
+    @default
+    bar: String
+}
+```
+
+
+### How many members used in output marked as required?
+
+As of March 17, 2021, 4,805 members.
+
+
+### Can we disallow the box trait on members in 1.0 right now?
+
+Not backward compatibly. For example, in AWS, it's used on 271 members.
+
+
+### How often has the `@required` trait been removed from members in AWS?
+
+The required trait has been removed from a structure member 618 different times
+across 105 different services. Of those removals, 75 targeted structures, 350
+targeted strings, and the remainder targeted lists, maps, and other simple
+shapes.
+
+
+### How often has the `@required` trait been added to members in AWS?
+
+The required trait has been added to a member 9 different times across 5
+different services.
+
+
+### Why not base backward compatibility rules around inputs and outputs?
+
+This proposal talks about backward compatibility of structure members and does
```

Comment:
I updated this section of the proposal to now special case input shapes as having more relaxed backward compatibility semantics. I'll do a followup design doc, but basically the `@input` trait will be used as a kind of contract that a structure is only ever used as the input of a single operation. Because of this knowledge, we can allow teams to remove the `@required` trait without replacing it with the `@default` trait, and code generators can account for that however necessary. There aren't any affordances made explicitly for output shapes, and it's not something the spec recommends, but you can implement more relaxed requirements there too if it's something you need.